squidpy/gr/_build.py

```diff
@@ -104,8 +119,74 @@ def spatial_neighbors(
     else:
         coord_type = CoordType(coord_type)
 
-    start = logg.info(f"Creating graph using `{coord_type}` coordinates and `{transform}` transform")
+    if library_key is not None:
+        _assert_categorical_obs(adata, key=library_key)
+        libs = adata.obs[library_key].cat.categories
+        unique_obs_names = make_index_unique(adata.obs_names)
+    else:
+        libs = [None]
+
+    start = logg.info(
+        f"Creating graph using `{coord_type}` coordinates and `{transform}` transform and `{len(libs)}` libraries."
+    )
+    _build_fun = partial(
+        _spatial_neighbor,
+        spatial_key=spatial_key,
+        coord_type=coord_type,
+        n_neighs=n_neighs,
+        radius=radius,
+        delaunay=delaunay,
+        n_rings=n_rings,
+        transform=transform,
+        set_diag=set_diag,
+    )
+
+    if library_key is not None:
+        adj_mats = []
+        dst_mats = []
+        obs_names = []
+        for lib in libs:
+            obs_names.extend(unique_obs_names[adata.obs[library_key] == lib])
+            Adj, Dst = _build_fun(adata[adata.obs[library_key] == lib])
+            adj_mats.append(Adj)
+            dst_mats.append(Dst)
+        adj_tmp = AnnData(block_diag(adj_mats, format="csr"), dtype="float64")
+        dst_tmp = AnnData(block_diag(dst_mats, format="csr"), dtype="float64")
+        adj_tmp.obs_names = dst_tmp.obs_names = adj_tmp.var_names = dst_tmp.var_names = obs_names
```

Comment:
Am wondering why use `AnnData` in this case, because of the amount of code it produces...
Would something like this work?
```python
ixs = []
for lib in libs:
    ixs.extend(np.where(adata.obs[library_key] == lib)[0])
    mats.append( _build_fun(adata[adata.obs[library_key] == lib]))
ixs = np.argsort(ixs)  # invert
Adj = block_diag(mats[0], format="csr")[ixs, :][:, ixs]
Dst = block_diag(mats[1], format="csr")[ixs, :][:, ixs]
```
That way you don't have to worry about unique indices, etc.

Reply:
Also gets rid of the pesky mixin mentioned below.

Reply:
that code block doesn't work, I think ok for now with AnnData, incorporated other comments

Reply:
> Can't {obs,var}_names be set in the AnnData constructor?

no